File: DotNetOnion/Cells/CellCerts.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using DotNetOnion.Helpers;

namespace DotNetOnion.Cells
{
    public class CellCerts : Cell
    {
        public List<Cert> Certs { get; set; } = new List<Cert>();
        public override byte Command => 129;

        public override void Deserialize(BinaryReader reader)
        {
            int certCount = reader.ReadByte();

            for (int i = 0; i < certCount; i++)
            {
                Certs.Add(new Cert
                {
                    Type = reader.ReadByte(),
                    Certificate = reader.ReadBytes(reader.ReadUInt16BigEndian())
                });
            }
        }

        public override void Serialize(BinaryWriter writer)
        {
            writer.Write((byte)Certs.Count);

            foreach (var cert in Certs)
            {
                writer.Write(cert.Type);
                writer.WriteBigEndian((ushort)cert.Certificate.Length);
                writer.Write(cert.Certificate);
            }
        }
    }

    public class Cert
    {
        public byte Type { get; set; }
        public byte[] Certificate { get; set; }
    }
}

File: NOnion/Cells/CellCerts.fs
namespace NOnion.Cells

open System.IO

open FSharpx.Collections

open NOnion.Utility.BinaryIO

type Cert =
    {
        Type: byte
        Certificate: array<byte>
    }

type CellCerts =
    {
        Certs: seq<Cert>
    }

    static member Deserialize(reader: BinaryReader) =

        let rec readCertificates certificates remainingCount =
            if remainingCount = 0 then
                certificates
            else
                let certificate =
                    {
                        Cert.Type = reader.ReadByte()
                        Cert.Certificate =
                            ReadBigEndianUInt16 reader
                            |> int
                            |> reader.ReadBytes
                    }

                readCertificates
                    (certificates @ [ certificate ])
                    (remainingCount - 1)

        let certificatesCount = reader.ReadByte() |> int
        let certs = readCertificates List.empty certificatesCount

        {
            Certs = certs
        }
        :> ICell

    interface ICell with

        member __.Command = 129uy

        member self.Serialize writer =

            let rec writeCertificates(certificates: seq<Cert>) =
                match Seq.tryHeadTail certificates with
                | None -> ()
                | Some(certificate, nextCertificates) ->
                    writer.Write certificate.Type

                    certificate.Certificate.Length
                    |> uint16
                    |> WriteUInt16BigEndian writer

                    writer.Write certificate.Certificate
                    nextCertificates |> writeCertificates

            self.Certs |> Seq.length |> uint8 |> writer.Write
            writeCertificates self.Certs
